### PR TITLE
Promote users to non-anonymous status even when linking account w/o pw

### LINKED ISSUE #1331: linkAndRetrieveDataWithCredential: doesn't correctly handle linking an anonymous account with emailed-link authentication
### Environment

  * Xcode version: 9.3.1
  * Firebase SDK version: 4.13.0 (but can see the same code in v5)
  * Firebase Component: Auth
  * Component version: 4.6.1

### Problem

If you create an anonymous user and then call `linkAndRetrieveDataWithCredential:` to link it with an emailed-link credential, it doesn't correctly set `_anonymous = NO` - because `updateEmail:password:callback:` is called, and since password is nil it skips that step. It's assuming that all `FIREmailPasswordAuthCredential` instances are email+password, whereas they could also be email+link.

#### Steps to reproduce:

1. Create an anonymous session
2. Go through the email+link authentication process
3. Call `linkAndRetrieveDataWithCredential:` to link the existing anonymous session with the new `FIREmailPasswordAuthCredential`
4. Call `[[[FIRAuth auth] currentUser] isAnonymous]` - it will return `YES` even though the user should no longer be anonymous 

#### Relevant Code:

```
[[FIRAuth auth]
        signInAnonymouslyWithCompletion:^(FIRUser *_Nullable user, NSError *_Nullable error) {
 
            FIRAuthCredential *credential = [FIREmailAuthProvider credentialWithEmail:email link:url.absoluteString];

            [[[FIRAuth auth] currentUser] linkWithCredential:credential completion:^(FIRUser *_Nullable user, NSError *_Nullable error) {
            
                   NSLog(@"Is new user still considered anonymous? %d", [[[FIRAuth auth] currentUser] isAnonymous]);

        }];
    }
];
```

### PR DESCRIPTION
**Pull request by morganchen12**
Fixes #1331.

TODO:
- [x] Unit tests

### CODE REVIEW

File: Firebase/Auth/Source/FIRUser.m [hunk 1]
```diff
@@ -959,6 +959,13 @@ - (void)linkAndRetrieveDataWithCredential:(FIRAuthCredential *)credential
         if (error) {
           callInMainThreadWithAuthDataResultAndError(completion, nil, error);
         } else {
+          // Mark users as non-anonymous when linking a credential even if they have
+          // an email but no password defined. This can happen if the user uses email
+          // link login instead of a password.
+          // This isn't a generalizable solution, and should be refactored later if other
+          // cases arise (in the refactoring, the updateEmail: method probably shouldn't
+          // also own account promotion responsibilities).
+          self->_anonymous = NO;
```
[the remark this answers is not in the document]
Comment (morganchen12) [reply Firebase/Auth/Source/FIRUser.m]: Done. Let me know if this is the right approach and I'll add tests.

**Comment by morganchen12** [comment]
@protocol86 tests have been added, PTAL when you get a chance.

Thanks!

File: Example/Auth/Tests/FIRUserTests.m [hunk 2]
```diff
@@ -1572,6 +1625,63 @@ - (void)testlinkAndRetrieveDataErrorAutoSignOut {
...
+    XCTAssertTrue(user.isAnonymous);
+
+    [user linkAndRetrieveDataWithCredential:linkEmailCredential
+                                 completion:^(FIRAuthDataResult *_Nullable
+                                              linkAuthResult,
+                                              NSError *_Nullable error) {
+      XCTAssertTrue([NSThread isMainThread]);
+      XCTAssertNil(error);
+      XCTAssertFalse(user.isAnonymous);
+      [expectation fulfill];
```
Comment (protocol86) [inline Example/Auth/Tests/FIRUserTests.m]: please assert the presence of the linked email as well.

File: Example/Auth/Tests/FIRUserTests.m [hunk 3]
```diff
@@ -583,6 +585,57 @@ - (void)testUpdateEmailSuccess {
   OCMVerifyAll(_mockBackend);
 }
 
+/** @fn testUpdateEmailWithAuthLinkAccountSuccess
+ @brief Tests a successful @c updateEmail:completion: call updates provider info.
```
Comment (protocol86) [inline Example/Auth/Tests/FIRUserTests.m]: Please fix indentation

File: Example/Auth/Tests/FIRUserTests.m [hunk 4]
```diff
@@ -1572,6 +1625,63 @@ - (void)testlinkAndRetrieveDataErrorAutoSignOut {
   OCMVerifyAll(_mockBackend);
 }
 
+/** @fn testLinkingAnonymousAccountsUpdatesIsAnonymous
+ @brief Tests the flow of a successful @c linkAndRetrieveDataWithCredential:completion:
```
Comment (protocol86) [inline Example/Auth/Tests/FIRUserTests.m]: Fix indentation.

File: Example/Auth/Tests/FIRUserTests.m [hunk 5]
```diff
@@ -2222,14 +2332,42 @@ - (void)getIDTokenResultForcingRefreshSuccessWithIDToken:(NSString *)idToken {
   OCMVerifyAll(_mockBackend);
 }
 
+/** @fn signInAnonymouslyWithMockGetAccountInfoResponse:completion:
+ @brief Signs in with an anonymous account with mocked backend end calls.
```
Comment (protocol86) [inline Example/Auth/Tests/FIRUserTests.m]: Fix indentation

File: Example/Auth/Tests/FIRUserTests.m [hunk 6]
```diff
@@ -2253,6 +2391,35 @@ - (void)signInWithEmailPasswordWithMockUserInfoResponse:(id)mockUserInfoResponse
   }];
 }
 
+/** @fn signInWithEmailLinkWithMockGetAccountInfoResponse:completion:
+ @brief Signs in with an email link auth account with mocked backend end calls.
```
Comment (protocol86) [inline Example/Auth/Tests/FIRUserTests.m]: Fix indentation.

**Comment by paulb777** [comment]
@morganchen12 Please add to the release note as well and we'll try to get this in M28.

File: Example/Auth/Tests/FIRUserTests.m (hunk 2 above)
Comment (morganchen12) [reply Example/Auth/Tests/FIRUserTests.m]: Done.

File: Example/Auth/Tests/FIRUserTests.m (hunk 3 above)
Comment (morganchen12) [reply Example/Auth/Tests/FIRUserTests.m]: Done.

File: Example/Auth/Tests/FIRUserTests.m (hunk 4 above)
Comment (morganchen12) [reply Example/Auth/Tests/FIRUserTests.m]: Done.

File: Example/Auth/Tests/FIRUserTests.m (hunk 5 above)
Comment (morganchen12) [reply Example/Auth/Tests/FIRUserTests.m]: Done.

File: Example/Auth/Tests/FIRUserTests.m (hunk 6 above)
Comment (morganchen12) [reply Example/Auth/Tests/FIRUserTests.m]: Done.

**Pull request merged**